### PR TITLE
Support Map type in mapAdd and mapSubtract

### PR DESCRIPTION
**Pull request by ildus**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

* Support Map type in `mapAdd` and `mapSubtract` functions
* Support (U)Int128, U(Int256) types in `mapAdd` and `mapSubtract` functions

### CODE REVIEW

File: docs/en/sql-reference/functions/tuple-map-functions.md [hunk 1]
```diff
@@ -70,23 +70,23 @@ Result:
...
+Arguments are [maps](../../sql-reference/data-types/map.md) or [tuples](../../sql-reference/data-types/tuple.md#tuplet1-t2) of two [arrays](../../sql-reference/data-types/array.md#data-type-array), where items in the first array represent keys, and the second array contains values for the each key. All key arrays should have same type, and all value arrays should contain items which are promote to the one type ([Int64](../../sql-reference/data-types/int-uint.md#int-ranges), [UInt64](../../sql-reference/data-types/int-uint.md#uint-ranges) or [Float64](../../sql-reference/data-types/float.md#float32-float64)). The common promoted type is used as a type for the result array.
 
 **Returned value**
 
--   Returns one [tuple](../../sql-reference/data-types/tuple.md#tuplet1-t2), where the first array contains the sorted keys and the second array contains values.
+-   Depending on the arguments returns one [map](../../sql-reference/data-types/map.md) o [tuple](../../sql-reference/data-types/tuple.md#tuplet1-t2), where the first array contains the sorted keys and the second array contains values.
```
Comment (Avogar) [inline docs/en/sql-reference/functions/tuple-map-functions.md:85]:
```suggestion
-   Depending on the arguments returns one [map](../../sql-reference/data-types/map.md) or [tuple](../../sql-reference/data-types/tuple.md#tuplet1-t2), where the first array contains the sorted keys and the second array contains values.
```

File: src/Functions/array/mapOp.cpp [hunk 2]
```diff
@@ -52,90 +56,150 @@ class FunctionMapOp : public IFunction
...
         return res;
     }
 
-    template <typename KeyType, bool is_str_key, typename ValType>
-    ColumnPtr execute2(size_t row_count, TupleMaps & args, const DataTypeTuple & res_type) const
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
     {
-        MutableColumnPtr res_tuple = res_type.createColumn();
+        if (arguments.size() < 2)
+            throw Exception(getName() + " accepts at least two maps", ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
```
Comment (Avogar) [inline src/Functions/array/mapOp.cpp:160]:
```suggestion
            throw Exception(getName() + " accepts at least two maps or map tuples", ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
```

File: src/Functions/array/mapOp.cpp [hunk 3]
```diff
@@ -147,28 +211,38 @@ class FunctionMapOp : public IFunction
...
+                        if (const auto * col_fixed = checkAndGetColumn<ColumnFixedString>(arg.key_column.get()))
+                            key = col_fixed->getDataAt(offset + j).toString();
+                        else if (const auto * col_str = checkAndGetColumn<ColumnString>(arg.key_column.get()))
+                            key = col_str->getDataAt(offset + j).toString();
+                        else
+                            // should not happen
+                            throw Exception(
+                                "Expected String or FixedString, got " + std::string(getTypeName(arg.key_column->getDataType()))
+                                    + " in " + getName(),
+                                ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
```
Comment (Avogar) [inline src/Functions/array/mapOp.cpp:237]: If it should not happen, use `ErrorCodes::LOGICAL_ERROR`

**Review by Avogar — APPROVED** [review]

**Comment by gyuton** [comment]
Internal documentation ticket: DOCSUP-11552.

**Pull request merged**